core/src/data_lookup/mod.rs

```diff
@@ -0,0 +1,234 @@
+use codec::{Decode, Encode, Input};
+use core::convert::TryFrom;
+use scale_info::{Type, TypeInfo};
+use sp_core::RuntimeDebug;
+use sp_std::{ops::Range, vec::Vec};
+use thiserror_no_std::Error;
+
+#[cfg(feature = "std")]
+use serde::{Deserialize, Serialize};
+
+use crate::{ensure, AppId};
+
+pub mod compact;
+use compact::CompactDataLookup;
+
+pub type DataLookupRange = Range<u32>;
+
+#[derive(Error, Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Error {
+	#[error("Input data is not sorted by AppId")]
+	DataNotSorted,
+	#[error("Data is empty on AppId {0}")]
+	DataEmptyOn(AppId),
+	#[error("Offset overflows")]
+	OffsetOverflows,
+}
+
+#[derive(PartialEq, Eq, Clone, Default, RuntimeDebug)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[cfg_attr(
+	feature = "std",
+	serde(try_from = "CompactDataLookup", into = "CompactDataLookup")
+)]
+pub struct DataLookup {
+	pub(crate) index: Vec<(AppId, DataLookupRange)>,
+}
+
+impl DataLookup {
+	pub fn len(&self) -> u32 {
+		self.index.last().map(|(_id, range)| range.end).unwrap_or(0)
+	}
+
+	pub fn is_empty(&self) -> bool {
+		self.len() == 0
+	}
+
+	pub fn range_of(&self, app_id: AppId) -> Option<DataLookupRange> {
+		self.index
+			.iter()
+			.find(|(id, _)| *id == app_id)
+			.map(|(_, range)| range)
+			.cloned()
+	}
+
+	pub fn projected_range_of(&self, app_id: AppId, chunk_size: u32) -> Option<DataLookupRange> {
+		self.range_of(app_id).and_then(|range| {
+			let start = range.start.checked_mul(chunk_size)?;
+			let end = range.end.checked_mul(chunk_size)?;
+			Some(start..end)
+		})
+	}
+
+	pub fn projected_ranges(&self, chunk_size: u32) -> Result<Vec<(AppId, Range<u32>)>, Error> {
+		self.index
+			.iter()
+			.map(|(id, range)| {
+				let start = range
+					.start
+					.checked_mul(chunk_size)
+					.ok_or(Error::OffsetOverflows)?;
+				let end = range
+					.end
+					.checked_mul(chunk_size)
+					.ok_or(Error::OffsetOverflows)?;
+				Ok((*id, start..end))
+			})
+			.collect()
+	}
+}
+
+impl DataLookup {
+	pub fn from_id_and_len_iter<I, A, L>(iter: I) -> Result<Self, Error>
```

Comment:
This is a though one. I would argue that using generics here is a wrong choice.

Looking at actual data that is being passed to this function we have the following cases:
- Vec<(u32, u32)>
- Vec<(u32, usize)>
- Vec<(AppId, u32)>
- Vec<(AppId, usize)>

Issues:
- On line 100 we try to convert len from type `L` to `u32`. This is error prone because the caller assumes that no constrains will be violated if we pass an object of type `usize`. Instead of allowing the user to get an error a better option will be to constrain this function ever more and set `L` to be of type `u32` instead of  the `TryFrom` . Reasoning: Why should the caller be punished for following the interface constraints.
- line 94 we convert id of type `A` to `AppId` which in unnecessary because in half of our cases we already pass AppId. This is confusing and make it harder to reason with this function. Instead of allowing the user to pass anything it would be much easier if we immediately specify that `A` is supposed to be `AppId` and let the caller deal with that. Reasoning: It's easier to understand and user this interface if we know what exactly type is needed and used.
- If for ergonomic reasons with want to still support passing both `AppId` and `u32` to this function, we can easily refactor out the content of `.map` to a new function and provide two interfaces for conversion. 1) `pub fn from_id_u32` 2) `pub fn from_AppId` 
- We could even make `L` to be type `NonZeroU32` which would make it more correct but less ergonomic

Reply:
I quite agree on this, but I choose generic to simplify code ( not needed extra `map(..)`) and dedup (it does not need `from_*` functions). 
The ergonomic reason simplifies quite well the testing tool.